accepted/00000-http-api.md

```diff
@@ -0,0 +1,214 @@
+
+
+# Detailed Design
+
+The design aims to provide a plain HTTP endpoint to the existing methods by reusing the available business logic with minimum redundancy. The bulk of the effort is to decouple the existing layers from protocol-dependent parts as much as possible, so that the most of the workflow can be utilized for both interfaces in an agnostic manner.
+
+![Current workflow](images/http-api-before.png)
+*Current workflow*
+
+The components marked with red are tightly coupled to the XMLRPC-specific functionality, and cannot be reused with HTTP as-is. In contrast, the green parts are agnostic, and can be used for both interfaces.
+
+The main goal is to restructure the workflow so that the agnostic parts can be isolated into separete components and be shared between two interfaces. This is especially important for handler classes, which are the linear growing components of the API implementation.
+
+To implement the plain HTTP over JSON interface, the existing [Spark framework](https://sparkjava.com) shall be used.
+
+![Proposed workflow](images/http-api-after.png)
+*Proposed workflow*
+
+The design is split into a number of areas that represent different aspects of the new API. The rest of the section describes each aspect in detail.
+
+1. [Endpoint registration](#1-endpoint_registration)
+2. [Authentication](#2-authentication)
+3. [I/O](#3-io)
+4. [Exception handling](#4-exception-handling)
+5. [Security](#5-security)
+6. [Documentation](#6-documentation)
+
+
+## 1. Endpoint Registration
+
+The most of the boilerplate code involves registration of the endpoint handlers, dispatching requests, and parameter matching. The XMLRPC framework provides an interface to register a class as a handler for a specific namespace. Further resolution of a method in a namespace is handled by the `BaseHandler` class using Java reflection to match the method names and parameters. `BaseHandler` is the base class that all the namespace handler classes inherit from.
+
+On the HTTP side, Spark allows registration of an endpoint as a "route" per each endpoint method, optionally including a number of named route parameters that construct the route path.
+
+The actual handler methods are designed to use POJOs as parameters. However, Spark expects a specific signature that takes `request` and `response` as parameters for route handlers. A simple wrapper shall unpack the parameters from the request and call the appropriate handler method for each request.
+
+A typical resource-oriented REST API relies on various HTTP verbs to provide contextual information on what type of operation an endpoint is used for. However, the XMLRPC protocol is designed around methods, rather than resources. Therefore, HTTP verbs will be restricted to GET for read-only, and POST for read-write operations regardless of whether they require a payload or not.
```

Comment:
Do we not use DELETE requests for removing things?

Reply:
The difference between a POST and a DELETE is only semantics. If we had a resource-oriented API, it would make sense to implement multiple verbs on the same endpoint like:
```
PUT /api/contentmanagement/project (create a project)
POST /api/contentmanagement/project (update a project)
DELETE /api/contentmanagement/project (delete a project)
```
But in our case, the action is already in the method name: `updateProject`, `deleteProject`, etc. so there's no real benefit in using alternative verbs.

Reply:
ok